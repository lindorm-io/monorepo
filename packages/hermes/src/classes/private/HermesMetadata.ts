/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { Dict } from "@lindorm/types";
import {
  MetaAggregate,
  MetaCommand,
  MetaEvent,
  MetaHandler,
  MetaQuery,
  MetaSaga,
  MetaTimeout,
  MetaView,
  WithHandlers,
} from "../../types";

type InternalArray =
  | "aggregates"
  | "commands"
  | "events"
  | "handlers"
  | "queries"
  | "sagas"
  | "timeouts"
  | "views";

export class HermesMetadata {
  private readonly aggregates: Array<MetaAggregate>;
  private readonly commands: Array<MetaCommand>;
  private readonly events: Array<MetaEvent>;
  private readonly handlers: Array<MetaHandler>;
  private readonly queries: Array<MetaQuery>;
  private readonly sagas: Array<MetaSaga>;
  private readonly timeouts: Array<MetaTimeout>;
  private readonly views: Array<MetaView>;

  public constructor() {
    this.aggregates = [];
    this.commands = [];
    this.events = [];
    this.handlers = [];
    this.queries = [];
    this.sagas = [];
    this.timeouts = [];
    this.views = [];
  }

  // add

  public addAggregate(metadata: MetaAggregate): void {
    if (this.aggregates.some((m) => m.target === metadata.target)) return;
    this.addMetadata("aggregates", metadata);
  }

  public addCommand(metadata: MetaCommand): void {
    if (this.commands.some((m) => m.target === metadata.target)) return;
    this.addMetadata("commands", metadata);
  }

  public addEvent(metadata: MetaEvent): void {
    if (this.events.some((m) => m.target === metadata.target)) return;
    this.addMetadata("events", metadata);
  }

  public addHandler(metadata: MetaHandler): void {
    if (
      this.handlers.some(
        (m) =>
          m.target === metadata.target &&
          m.trigger === metadata.trigger &&
          m.decorator === metadata.decorator,
      )
    ) {
      return;
    }
    this.addMetadata("handlers", metadata);
  }

  public addQuery(metadata: MetaQuery): void {
    if (this.queries.some((m) => m.target === metadata.target)) return;
    this.addMetadata("queries", metadata);
  }

  public addSaga(metadata: MetaSaga): void {
    if (this.sagas.some((m) => m.target === metadata.target)) return;
    this.addMetadata("sagas", metadata);
  }

  public addTimeout(metadata: MetaTimeout): void {
    if (this.timeouts.some((m) => m.target === metadata.target)) return;
    this.addMetadata("timeouts", metadata);
  }

  public addView(metadata: MetaView): void {
    if (this.views.some((m) => m.target === metadata.target)) return;
    this.addMetadata("views", metadata);
  }

  // get dtos

  public getCommand(Ctor: Function): MetaCommand | undefined {
    const [command] = this.getMeta<MetaCommand>(Ctor, "commands");
    if (!command) return;
    return command;
  }

  public getEvent(Ctor: Function): MetaEvent | undefined {
    const [event] = this.getMeta<MetaEvent>(Ctor, "events");
    if (!event) return;
    return event;
  }

  public getQuery(Ctor: Function): MetaQuery | undefined {
    const [query] = this.getMeta<MetaQuery>(Ctor, "queries");
    if (!query) return;
    return query;
  }

  public getTimeout(Ctor: Function): MetaTimeout | undefined {
    const [timeout] = this.getMeta<MetaTimeout>(Ctor, "timeouts");
    if (!timeout) return;
    return timeout;
  }

  // get models

  public getAggregate(Ctor: Function): WithHandlers<MetaAggregate> | undefined {
    return this.getWithHandlers<MetaAggregate>(Ctor, "aggregates");
  }

  public getSaga(Ctor: Function): WithHandlers<MetaSaga> | undefined {
    return this.getWithHandlers<MetaSaga>(Ctor, "sagas");
  }

  public getView(Ctor: Function): WithHandlers<MetaView> | undefined {
    return this.getWithHandlers<MetaView>(Ctor, "views");
  }

  // find

  public findCommand(name: string): MetaCommand | undefined {
    return this.commands.find((meta) => meta.name === name);
  }

  public findEvent(name: string, version: number): MetaEvent | undefined {
    return this.events.find((meta) => meta.name === name && meta.version === version);
  }

  public findQuery(name: string): MetaQuery | undefined {
    return this.queries.find((meta) => meta.name === name);
  }

  public findTimeout(name: string): MetaTimeout | undefined {
    return this.timeouts.find((meta) => meta.name === name);
  }

  // private

  private addMetadata(key: InternalArray, metadata: any): void {
    this[key].push(metadata);
  }

  private getMeta<T>(target: Function, arr: InternalArray): Array<T> {
    const collected: Array<any> = [];

    let current: any = target;

    while (current && current !== Function.prototype) {
      collected.push(...this[arr].filter((meta: any) => meta.target === current));
      current = Object.getPrototypeOf(current);
    }

    return collected;
  }

  private getWithHandlers<T extends Dict>(
    Ctor: Function,
    arr: InternalArray,
  ): WithHandlers<T> | undefined {
    const [item] = this.getMeta<T>(Ctor, arr);

    if (!item) return;

    const handlers = this.getMeta<MetaHandler>(Ctor, "handlers");

    return { ...item, handlers };
  }
}
