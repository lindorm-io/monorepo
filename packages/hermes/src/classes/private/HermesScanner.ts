import { isObject, isString } from "@lindorm/is";
import { IScanData, Scanner } from "@lindorm/scanner";
import { Constructor, Dict } from "@lindorm/types";
import { HermesScannerError } from "../../errors";
import { HermesScannerInput } from "../../types";

export class HermesScanner {
  public static scan<T extends Dict = Dict>(
    input: HermesScannerInput<T>,
  ): Array<Constructor<T>> {
    const classes = input.filter(
      (a) => !isObject(a) && !isString(a) && (a as T).prototype,
    ) as Array<Constructor<T>>;

    const strings = input.filter((a) => isString(a));

    const result: Array<Constructor<T>> = [...classes];

    if (!strings.length) return result;

    for (const path of strings) {
      const item = HermesScanner.scanner.scan(path);

      if (item.isDirectory) {
        result.push(...HermesScanner.scanDirectory<T>(item));
      }
      if (item.isFile) {
        result.push(HermesScanner.scanFile<T>(item));
      }
    }

    return result;
  }

  // private

  private static scanDirectory<T extends Dict = Dict>(
    data: IScanData,
  ): Array<Constructor<T>> {
    const result: Array<Constructor<T>> = [];

    for (const child of data.children) {
      if (child.isDirectory) {
        result.push(...HermesScanner.scanDirectory<T>(child));
      }
      if (child.isFile) {
        result.push(HermesScanner.scanFile<T>(child));
      }
    }

    return result;
  }

  private static scanFile<T extends Dict = Dict>(data: IScanData): Constructor<T> {
    const module = HermesScanner.scanner.require<Constructor<T>>(data.fullPath);

    const entries = Object.entries(module);
    if (entries.length === 0) {
      throw new HermesScannerError(`No classes found in file: ${data.fullPath}`);
    }

    let result: Constructor<T> | null = null;

    for (const value of Object.values(module)) {
      if (result) break;

      if (value.prototype) {
        result = value;
        break;
      }
    }

    if (!result) {
      throw new HermesScannerError(`No classes found in file: ${data.fullPath}`);
    }

    return result as Constructor<T>;
  }

  private static get scanner(): Scanner {
    return new Scanner({
      deniedFilenames: [/^index$/],
      deniedTypes: [/^fixture$/, /^spec$/, /^test$/, /^integration$/],
    });
  }
}
