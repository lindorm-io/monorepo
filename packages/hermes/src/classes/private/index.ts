export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
