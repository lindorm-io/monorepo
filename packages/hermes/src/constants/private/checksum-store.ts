import { ChecksumStoreAttributes, StoreIndexes } from "../../types";

export const CHECKSUM_STORE = "checksum_store";
export const CHECKSUM_STORE_INDEXES: StoreIndexes<ChecksumStoreAttributes> = [
  {
    fields: ["id", "name", "context", "event_id"],
    name: "cst_pkey",
    unique: true,
  },
  {
    fields: ["id", "name", "context"],
    name: "cst_idx",
    unique: false,
  },
];
