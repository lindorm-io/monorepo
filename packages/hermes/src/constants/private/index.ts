export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
