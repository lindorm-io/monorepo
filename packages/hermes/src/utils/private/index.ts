export * from "./assert-checksum";
export * from "./compose-metadata";
export * from "./create-checksum";
export * from "./extract-data-transfer-object";
export * from "./get-view-store-name";
export * from "./global";
export * from "./recover-dto";
export * from "./recover-error";
export * from "./verify-name";
