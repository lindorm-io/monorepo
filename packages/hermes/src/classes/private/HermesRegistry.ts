import { snakeCase } from "@lindorm/case";
import { Constructor, Dict } from "@lindorm/types";
import {
  HermesAggregateCommandHandler,
  HermesAggregateErrorHandler,
  HermesAggregateEventHandler,
  HermesSagaErrorHandler,
  HermesSagaEventHandler,
  HermesSagaIdHandler,
  HermesSagaTimeoutHandler,
  HermesViewErrorHandler,
  HermesViewEventHandler,
  HermesViewIdHandler,
  HermesViewQueryHandler,
} from "../../handlers";
import {
  IAggregateCommandHandler,
  IAggregateErrorHandler,
  IAggregateEventHandler,
  IHermesRegistry,
  ISagaErrorHandler,
  ISagaEventHandler,
  ISagaIdHandler,
  ISagaTimeoutHandler,
  IViewErrorHandler,
  IViewEventHandler,
  IViewIdHandler,
  IViewQueryHandler,
} from "../../interfaces";
import {
  AggregateCommandCallback,
  AggregateErrorCallback,
  AggregateEventCallback,
  HermesRegistryOptions,
  HermesScannerInput,
  MetaAggregate,
  MetaCommand,
  MetaEvent,
  MetaHandler,
  MetaQuery,
  MetaSaga,
  MetaTimeout,
  MetaView,
  RegistryAggregate,
  RegistryCommand,
  RegistryEvent,
  RegistryQuery,
  RegistrySaga,
  RegistryTimeout,
  RegistryView,
  SagaErrorCallback,
  SagaEventCallback,
  SagaIdCallback,
  SagaTimeoutCallback,
  ViewErrorCallback,
  ViewEventCallback,
  ViewIdCallback,
  ViewQueryCallback,
} from "../../types";
import { globalHermesMetadata } from "../../utils/private";
import { HermesScanner } from "./HermesScanner";

export class HermesRegistry implements IHermesRegistry {
  private readonly namespace: string;

  private readonly commandCtors: Array<Constructor>;
  private readonly eventCtors: Array<Constructor>;
  private readonly queryCtors: Array<Constructor>;
  private readonly timeoutCtors: Array<Constructor>;

  public readonly commands: Array<RegistryCommand>;
  public readonly events: Array<RegistryEvent>;
  public readonly queries: Array<RegistryQuery>;
  public readonly timeouts: Array<RegistryTimeout>;

  public readonly aggregates: Array<RegistryAggregate>;
  public readonly sagas: Array<RegistrySaga>;
  public readonly views: Array<RegistryView>;

  public readonly aggregateCommandHandlers: Array<IAggregateCommandHandler>;
  public readonly aggregateErrorHandlers: Array<IAggregateErrorHandler>;
  public readonly aggregateEventHandlers: Array<IAggregateEventHandler>;

  public readonly sagaErrorHandlers: Array<ISagaErrorHandler>;
  public readonly sagaEventHandlers: Array<ISagaEventHandler>;
  public readonly sagaIdHandlers: Array<ISagaIdHandler>;
  public readonly sagaTimeoutHandlers: Array<ISagaTimeoutHandler>;

  public readonly viewErrorHandlers: Array<IViewErrorHandler>;
  public readonly viewEventHandlers: Array<IViewEventHandler>;
  public readonly viewIdHandlers: Array<IViewIdHandler>;
  public readonly viewQueryHandlers: Array<IViewQueryHandler>;

  public constructor(options: HermesRegistryOptions = {}) {
    this.namespace = options.namespace ?? "hermes";

    this.commandCtors = [];
    this.eventCtors = [];
    this.queryCtors = [];
    this.timeoutCtors = [];

    this.commands = [];
    this.events = [];
    this.queries = [];
    this.timeouts = [];

    this.aggregates = [];
    this.sagas = [];
    this.views = [];

    this.aggregateCommandHandlers = [];
    this.aggregateErrorHandlers = [];
    this.aggregateEventHandlers = [];

    this.sagaIdHandlers = [];
    this.sagaErrorHandlers = [];
    this.sagaEventHandlers = [];
    this.sagaTimeoutHandlers = [];

    this.viewIdHandlers = [];
    this.viewErrorHandlers = [];
    this.viewEventHandlers = [];
    this.viewQueryHandlers = [];
  }

  // add

  public add(input: HermesScannerInput): void {
    this.addCommands(input);
    this.addEvents(input);
    this.addQueries(input);
    this.addTimeouts(input);

    this.addAggregates(input);
    this.addSagas(input);
    this.addViews(input);
  }

  // add dtos

  public addCommands(input: HermesScannerInput): void {
    const commands = HermesScanner.scan(input);

    for (const Command of commands) {
      if (this.commands.some((c) => c.target === Command)) continue;

      const metadata = globalHermesMetadata.getCommand(Command);
      if (!metadata) continue;

      this.commandCtors.push(Command);
    }
  }

  public addEvents(input: HermesScannerInput): void {
    const events = HermesScanner.scan(input);

    for (const Event of events) {
      if (this.events.some((c) => c.target === Event)) continue;

      const metadata = globalHermesMetadata.getEvent(Event);
      if (!metadata) continue;

      this.eventCtors.push(Event);
    }
  }

  public addQueries(input: HermesScannerInput): void {
    const queries = HermesScanner.scan(input);

    for (const Query of queries) {
      if (this.queries.some((c) => c.target === Query)) continue;

      const metadata = globalHermesMetadata.getQuery(Query);
      if (!metadata) continue;

      this.queryCtors.push(Query);
    }
  }

  public addTimeouts(input: HermesScannerInput): void {
    const timeouts = HermesScanner.scan(input);

    for (const Timeout of timeouts) {
      if (this.timeouts.some((c) => c.target === Timeout)) continue;

      const metadata = globalHermesMetadata.getTimeout(Timeout);
      if (!metadata) continue;

      this.timeoutCtors.push(Timeout);
    }
  }

  // add handlers

  public addAggregates(input: HermesScannerInput): void {
    const aggregates = HermesScanner.scan(input);

    for (const Aggregate of aggregates) {
      if (this.aggregates.some((c) => c.target === Aggregate)) continue;

      const metadata = globalHermesMetadata.getAggregate(Aggregate);
      if (!metadata) return;

      for (const handler of metadata.handlers) {
        switch (handler.decorator) {
          case "AggregateCommandHandler":
            this.registerAggregateCommandHandler(metadata, handler);
            break;

          case "AggregateErrorHandler":
            this.registerAggregateErrorHandler(metadata, handler);
            break;

          case "AggregateEventHandler":
            this.registerAggregateEventHandler(metadata, handler);
            break;

          default:
            break;
        }
      }

      this.aggregates.push({
        encryption: metadata.encryption,
        name: metadata.name,
        namespace: metadata.namespace ?? this.namespace,
        target: Aggregate,
      });
    }
  }

  public addSagas(input: HermesScannerInput): void {
    const sagas = HermesScanner.scan(input);

    const aggregates: Array<MetaAggregate> = [];

    for (const Saga of sagas) {
      if (this.sagas.some((c) => c.target === Saga)) continue;

      const saga = globalHermesMetadata.getSaga(Saga);
      if (!saga) continue;

      for (const Aggregate of saga.aggregates) {
        const aggregate = globalHermesMetadata.getAggregate(Aggregate);

        if (!aggregate) {
          throw new Error(`Aggregate not found for saga: ${saga.name}`);
        }

        aggregates.push(aggregate);

        for (const handler of saga.handlers) {
          switch (handler.decorator) {
            case "SagaErrorHandler":
              this.registerSagaErrorHandler(aggregate, saga, handler);
              break;

            case "SagaEventHandler":
              this.registerSagaEventHandler(aggregate, saga, handler);
              break;

            case "SagaIdHandler":
              this.registerSagaIdHandler(aggregate, saga, handler);
              break;

            case "SagaTimeoutHandler":
              this.registerSagaTimeoutHandler(aggregate, saga, handler);
              break;

            default:
              break;
          }
        }
      }

      this.sagas.push({
        aggregates: aggregates.map((a) => ({
          name: a.name,
          namespace: a.namespace ?? this.namespace,
        })),
        name: saga.name,
        namespace: saga.namespace ?? this.namespace,
        target: Saga,
      });
    }
  }

  public addViews(input: HermesScannerInput): void {
    const views = HermesScanner.scan(input);

    const aggregates: Array<MetaAggregate> = [];

    for (const View of views) {
      if (this.views.some((c) => c.target === View)) continue;

      const view = globalHermesMetadata.getView(View);
      if (!view) continue;

      for (const Aggregate of view.aggregates) {
        const aggregate = globalHermesMetadata.getAggregate(Aggregate);

        if (!aggregate) {
          throw new Error(`Aggregate not found for view: ${view.name}`);
        }

        aggregates.push(aggregate);

        for (const handler of view.handlers) {
          switch (handler.decorator) {
            case "ViewErrorHandler":
              this.registerViewErrorHandler(aggregate, view, handler);
              break;

            case "ViewEventHandler":
              this.registerViewEventHandler(aggregate, view, handler);
              break;

            case "ViewIdHandler":
              this.registerViewIdHandler(aggregate, view, handler);
              break;

            case "ViewQueryHandler":
              this.registerViewQueryHandler(view, handler);
              break;

            default:
              break;
          }
        }
      }

      this.views.push({
        aggregates: aggregates.map((a) => ({
          name: a.name,
          namespace: a.namespace ?? this.namespace,
        })),
        name: view.name,
        namespace: view.namespace ?? this.namespace,
        source: view.source,
        target: View,
      });
    }
  }

  // get dto

  public getCommand(Command: Constructor): RegistryCommand {
    const command = this.commands.find((c) => c.target === Command);
    if (!command) {
      throw new Error(`Command not found: ${Command.name}`);
    }
    return command;
  }

  public getEvent(Event: Constructor): RegistryEvent {
    const event = this.events.find((c) => c.target === Event);
    if (!event) {
      throw new Error(`Event not found: ${Event.name}`);
    }
    return event;
  }

  public getQuery(Query: Constructor): RegistryQuery {
    const query = this.queries.find((c) => c.target === Query);
    if (!query) {
      throw new Error(`Query not found: ${Query.name}`);
    }
    return query;
  }

  public getTimeout(Timeout: Constructor): RegistryTimeout {
    const timeout = this.timeouts.find((c) => c.target === Timeout);
    if (!timeout) {
      throw new Error(`Timeout not found: ${Timeout.name}`);
    }
    return timeout;
  }

  // is dto

  public isCommand(Command: Constructor): boolean {
    return !!this.commands.find((c) => c.target === Command);
  }

  public isEvent(Event: Constructor): boolean {
    return !!this.events.find((c) => c.target === Event);
  }

  public isQuery(Query: Constructor): boolean {
    return !!this.queries.find((c) => c.target === Query);
  }

  public isTimeout(Timeout: Constructor): boolean {
    return !!this.timeouts.find((c) => c.target === Timeout);
  }

  // private dtos

  private registerCommand(command: MetaCommand, aggregate: MetaAggregate): void {
    if (!this.commandCtors.includes(command.target)) {
      throw new Error(`Command not found: ${command.target.name}`);
    }

    if (this.commands.some((c) => c.target === command.target)) {
      return;
    }

    this.commands.push({
      aggregate: {
        name: aggregate.name,
        namespace: aggregate.namespace ?? this.namespace,
      },
      name: command.name,
      target: command.target,
    });
  }

  private registerEvent(event: MetaEvent, aggregate: MetaAggregate): void {
    if (!this.eventCtors.includes(event.target)) {
      throw new Error(`Event not found: ${event.target.name}`);
    }

    if (this.events.some((c) => c.target === event.target)) {
      return;
    }

    this.events.push({
      aggregate: {
        name: aggregate.name,
        namespace: aggregate.namespace ?? this.namespace,
      },
      name: event.name,
      target: event.target,
      version: event.version,
    });
  }

  private registerQuery(query: MetaQuery, view: MetaView): void {
    if (!this.queryCtors.includes(query.target)) {
      throw new Error(`Query not found: ${query.target.name}`);
    }

    if (this.queries.some((c) => c.target === query.target)) {
      return;
    }

    this.queries.push({
      name: query.name,
      target: query.target,
      view: { name: view.name, namespace: view.namespace ?? this.namespace },
    });
  }

  private registerTimeout(
    timeout: MetaTimeout,
    aggregate: MetaAggregate,
    saga: MetaSaga,
  ): void {
    if (!this.timeoutCtors.includes(timeout.target)) {
      throw new Error(`Timeout not found: ${timeout.target.name}`);
    }

    if (this.timeouts.some((c) => c.target === timeout.target)) {
      return;
    }

    this.timeouts.push({
      aggregate: {
        name: aggregate.name,
        namespace: aggregate.namespace ?? this.namespace,
      },
      name: timeout.name,
      saga: { name: saga.name, namespace: saga.namespace ?? this.namespace },
      target: timeout.target,
    });
  }

  // private aggregate handlers

  private registerAggregateCommandHandler(
    aggregate: MetaAggregate,
    handler: MetaHandler<AggregateCommandCallback<Constructor, Dict>>,
  ): void {
    const command = globalHermesMetadata.getCommand(handler.trigger);

    if (!command) {
      throw new Error(`Command not found for handler: ${handler.key}`);
    }

    if (command.target !== handler.trigger) {
      throw new Error(`Command/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.aggregateCommandHandlers.some(
        (h) =>
          h.command === command.name &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace,
      )
    ) {
      return;
    }

    this.registerCommand(command, aggregate);

    this.aggregateCommandHandlers.push(
      new HermesAggregateCommandHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        command: command.name,
        conditions: handler.conditions ?? undefined,
        encryption: handler.encryption ?? aggregate.encryption ?? false,
        key: handler.key,
        schema: handler.schema ?? undefined,
        handler: handler.handler,
      }),
    );
  }

  private registerAggregateErrorHandler(
    aggregate: MetaAggregate,
    handler: MetaHandler<AggregateErrorCallback<Constructor>>,
  ): void {
    if (
      this.viewErrorHandlers.some(
        (h) =>
          h.error === handler.trigger.name &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace,
      )
    ) {
      return;
    }

    this.aggregateErrorHandlers.push(
      new HermesAggregateErrorHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        error: snakeCase(handler.trigger.name),
        key: handler.key,
        handler: handler.handler,
      }),
    );
  }

  private registerAggregateEventHandler(
    aggregate: MetaAggregate,
    handler: MetaHandler<AggregateEventCallback<Constructor, Dict>>,
  ): void {
    const event = globalHermesMetadata.getEvent(handler.trigger);

    if (!event) {
      throw new Error(`Event not found for handler: ${handler.key}`);
    }

    if (event.target !== handler.trigger) {
      throw new Error(`Event/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.aggregateEventHandlers.some(
        (h) =>
          h.event.name === event.name &&
          h.event.version === event.version &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace,
      )
    ) {
      return;
    }

    this.registerEvent(event, aggregate);

    this.aggregateEventHandlers.push(
      new HermesAggregateEventHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        event: { name: event.name, version: event.version },
        key: handler.key,
        handler: handler.handler,
      }),
    );
  }

  // private saga handlers

  private registerSagaErrorHandler(
    aggregate: MetaAggregate,
    saga: MetaSaga,
    handler: MetaHandler<SagaErrorCallback<Constructor>>,
  ): void {
    if (
      this.sagaErrorHandlers.some(
        (h) =>
          h.error === handler.trigger.name &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.saga.name === saga.name &&
          h.saga.namespace === saga.namespace,
      )
    ) {
      return;
    }

    this.sagaErrorHandlers.push(
      new HermesSagaErrorHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        error: snakeCase(handler.trigger.name),
        key: handler.key,
        saga: { name: saga.name, namespace: saga.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerSagaEventHandler(
    aggregate: MetaAggregate,
    saga: MetaSaga,
    handler: MetaHandler<SagaEventCallback<Constructor, Dict>>,
  ): void {
    const event = globalHermesMetadata.getEvent(handler.trigger);

    if (!event) {
      throw new Error(`Event not found for handler: ${handler.key}`);
    }

    if (event.target !== handler.trigger) {
      throw new Error(`Event/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.sagaEventHandlers.some(
        (h) =>
          h.event.name === event.name &&
          h.event.version === event.version &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.saga.name === saga.name &&
          h.saga.namespace === saga.namespace,
      )
    ) {
      return;
    }

    this.registerEvent(event, aggregate);

    this.sagaEventHandlers.push(
      new HermesSagaEventHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        conditions: handler.conditions ?? undefined,
        event: { name: event.name, version: event.version },
        key: handler.key,
        saga: { name: saga.name, namespace: saga.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerSagaIdHandler(
    aggregate: MetaAggregate,
    saga: MetaSaga,
    handler: MetaHandler<SagaIdCallback<Constructor>>,
  ): void {
    const event = globalHermesMetadata.getEvent(handler.trigger);

    if (!event) {
      throw new Error(`Event not found for handler: ${handler.key}`);
    }

    if (event.target !== handler.trigger) {
      throw new Error(`Event/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.sagaIdHandlers.some(
        (h) =>
          h.event.name === event.name &&
          h.event.version === event.version &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.saga.name === saga.name &&
          h.saga.namespace === saga.namespace,
      )
    ) {
      return;
    }

    this.registerEvent(event, aggregate);

    this.sagaIdHandlers.push(
      new HermesSagaIdHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        event: { name: event.name, version: event.version },
        key: handler.key,
        saga: { name: saga.name, namespace: saga.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerSagaTimeoutHandler(
    aggregate: MetaAggregate,
    saga: MetaSaga,
    handler: MetaHandler<SagaTimeoutCallback<Constructor, Dict>>,
  ): void {
    const timeout = globalHermesMetadata.getTimeout(handler.trigger);

    if (!timeout) {
      throw new Error(`Timeout not found for handler: ${handler.key}`);
    }

    if (timeout.target !== handler.trigger) {
      throw new Error(`Timeout/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.sagaTimeoutHandlers.some(
        (h) =>
          h.timeout === timeout.name &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.saga.name === saga.name &&
          h.saga.namespace === saga.namespace,
      )
    ) {
      return;
    }

    this.registerTimeout(timeout, aggregate, saga);

    this.sagaTimeoutHandlers.push(
      new HermesSagaTimeoutHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        key: handler.key,
        saga: { name: saga.name, namespace: saga.namespace ?? this.namespace },
        timeout: timeout.name,
        handler: handler.handler,
      }),
    );
  }

  // private view handlers

  private registerViewErrorHandler(
    aggregate: MetaAggregate,
    view: MetaView,
    handler: MetaHandler<ViewErrorCallback<Constructor>>,
  ): void {
    if (
      this.viewErrorHandlers.some(
        (h) =>
          h.error === handler.trigger.name &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.view.name === view.name &&
          h.view.namespace === view.namespace,
      )
    ) {
      return;
    }

    this.viewErrorHandlers.push(
      new HermesViewErrorHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        error: snakeCase(handler.trigger.name),
        key: handler.key,
        view: { name: view.name, namespace: view.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerViewEventHandler(
    aggregate: MetaAggregate,
    view: MetaView,
    handler: MetaHandler<ViewEventCallback<Constructor, Dict>>,
  ): void {
    const event = globalHermesMetadata.getEvent(handler.trigger);

    if (!event) {
      throw new Error(`Event not found for handler: ${handler.key}`);
    }

    if (event.target !== handler.trigger) {
      throw new Error(`Event/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.viewEventHandlers.some(
        (h) =>
          h.event.name === event.name &&
          h.event.version === event.version &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.view.name === view.name &&
          h.view.namespace === view.namespace,
      )
    ) {
      return;
    }

    this.registerEvent(event, aggregate);

    this.viewEventHandlers.push(
      new HermesViewEventHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        conditions: handler.conditions ?? undefined,
        event: { name: event.name, version: event.version },
        key: handler.key,
        source: view.source,
        view: { name: view.name, namespace: view.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerViewIdHandler(
    aggregate: MetaAggregate,
    view: MetaView,
    handler: MetaHandler<ViewIdCallback<Constructor>>,
  ): void {
    const event = globalHermesMetadata.getEvent(handler.trigger);

    if (!event) {
      throw new Error(`Event not found for handler: ${handler.key}`);
    }

    if (event.target !== handler.trigger) {
      throw new Error(`Event/Aggregate mismatch for handler: ${handler.key}`);
    }

    if (
      this.viewIdHandlers.some(
        (h) =>
          h.event.name === event.name &&
          h.event.version === event.version &&
          h.aggregate.name === aggregate.name &&
          h.aggregate.namespace === aggregate.namespace &&
          h.view.name === view.name &&
          h.view.namespace === view.namespace,
      )
    ) {
      return;
    }

    this.registerEvent(event, aggregate);

    this.viewIdHandlers.push(
      new HermesViewIdHandler({
        aggregate: {
          name: aggregate.name,
          namespace: aggregate.namespace ?? this.namespace,
        },
        event: { name: event.name, version: event.version },
        key: handler.key,
        view: { name: view.name, namespace: view.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }

  private registerViewQueryHandler(
    view: MetaView,
    handler: MetaHandler<ViewQueryCallback<Constructor, Dict>>,
  ): void {
    const query = globalHermesMetadata.getQuery(handler.trigger);

    if (!query) {
      throw new Error(`Query not found for handler: ${handler.key}`);
    }

    if (
      this.viewQueryHandlers.some(
        (h) =>
          h.query === query.name &&
          h.source === view.source &&
          h.view.name === view.name &&
          h.view.namespace === view.namespace,
      )
    ) {
      return;
    }

    this.registerQuery(query, view);

    this.viewQueryHandlers.push(
      new HermesViewQueryHandler({
        key: handler.key,
        query: query.name,
        source: view.source,
        view: { name: view.name, namespace: view.namespace ?? this.namespace },
        handler: handler.handler,
      }),
    );
  }
}
