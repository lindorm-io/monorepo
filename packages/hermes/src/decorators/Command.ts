import { Constructor } from "@lindorm/types";
import { CommandDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Command(options?: CommandDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addCommand({
      name: options?.name || name,
      target: target as unknown as Constructor,
    });
  };
}
