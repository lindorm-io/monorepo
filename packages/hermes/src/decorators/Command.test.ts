import { globalHermesMetadata } from "../utils/private";
import { Command } from "./Command";

describe("Command Decorator", () => {
  test("should add metadata", () => {
    @Command()
    class TestCommand {}

    expect(globalHermesMetadata.getCommand(TestCommand)).toMatchSnapshot();
  });

  test("should add metadata with custom options", () => {
    @Command({ name: "custom_command" })
    class TestCommandOptions {}

    expect(globalHermesMetadata.getCommand(TestCommandOptions)).toMatchSnapshot();
  });
});
