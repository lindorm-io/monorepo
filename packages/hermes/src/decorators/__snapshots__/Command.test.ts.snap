// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Decorator should add metadata 1`] = `
{
  "name": "test_command",
  "target": [Function],
}
`;

exports[`Command Decorator should add metadata with custom options 1`] = `
{
  "name": "custom_command",
  "target": [Function],
}
`;
